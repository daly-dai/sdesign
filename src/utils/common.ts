/**
 * 生成唯一code
 * @param num code的位数
 * @returns
 */
export function createCode(num = 4) {
  const UP_ALPHABET = [
    'A',
    'B',
    'C',
    'D',
    'E',
    'F',
    'G',
    'H',
    'I',
    'J',
    'K',
    'L',
    'M',
    'N',
    'O',
    'P',
    'Q',
    'R',
    'S',
    'T',
    'U',
    'V',
    'W',
    'X',
    'Y',
    'Z',
  ];
  const DOWN_ALPHABET = [
    'a',
    'b',
    'c',
    'd',
    'e',
    'f',
    'g',
    'h',
    'i',
    'j',
    'k',
    'l',
    'm',
    'n',
    'o',
    'p',
    'q',
    'r',
    's',
    't',
    'u',
    'v',
    'w',
    'x',
    'y',
    'z',
  ];
  const NUMBER = ['1', '2', '3', '4', '5', '6', '7', '8', '9']; //用数组chars保存验证码里出现的字符

  const chars = UP_ALPHABET.concat(DOWN_ALPHABET, NUMBER);

  let randCode = ''; //定义一个初始值为空的字符串变量为最终产生的随机验证码

  for (let i = 0; i < num; i++) {
    //0-1的随机小数 -->  0~数组长度-1的范围   取整
    const randPosition = Math.floor(Math.random() * (chars.length - 1)); //每次生成一个随机数的位置

    randCode += chars[randPosition]; //带有随机位置作为下标，指示到当前随机产生的某个字符上，将其连接到随机验证码的后面
  }

  return randCode;
}

/**
 * 处理文件名称,超过限制字数保留前三后三
 * @param name
 * @returns
 */
export function dispatchFileName(name: string, nameLimit: number) {
  if (!name) return '';

  let FileDataType = '';

  if (name.includes('.')) {
    const nameList = name.split('.');

    if (nameList?.length > 1) {
      FileDataType = `.${nameList[nameList.length - 1]}`;
    }
  }

  let fileName = name;

  if (FileDataType) {
    fileName = name.replace(FileDataType, '');
  }

  if (fileName.length > nameLimit) {
    const lastThreeChars = fileName.substring(fileName.length - 3); // 截取后三个字符
    const firstThreeChars = fileName.substring(0, 3); // 截取前三个字符

    return `${firstThreeChars}...${lastThreeChars}${FileDataType}`;
  }

  return name;
}

/**
 * @description 创建数组
 * @param num
 * @returns
 */
export function createArray(num = 4) {
  return Array.from({ length: num }, (_, i) => i);
}

/**
 * 获取文件base64相关数据
 * @param file
 * @returns
 */
export function getBase64(file: any): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = (error) => reject(error);
  });
}
/**
 * 动态引入js
 *
 * @param {string} src - js链接地址
 * @example
 *
 * ```ts
 * import { injectScript } from 'super-tools-lib'
 *
 * injectScript(src)
 * ```
 */
export function injectScript(src: string) {
  const s = document.createElement('script');
  s.type = 'text/javascript';
  s.async = true;
  s.src = src;
  const t = document.getElementsByTagName('script')[0] as any;

  t.parentNode.insertBefore(s, t);
}

/**
 * 复制文本到剪切板
 * @param text 需要复制的文本
 */
export async function copyToClipboard(
  text: string,
  successCb?: () => void,
  errorCb?: () => void,
): Promise<void> {
  // 检查权限的示例函数，具体实现取决于你的框架和API
  async function checkPermission(): Promise<void> {
    if (!('clipboardWrite' in navigator.clipboard)) {
      throw new Error('Clipboard API is not supported.');
    }
  }

  try {
    // 检查是否具有剪贴板写入权限的函数
    await checkPermission(); // 检查权限，根据实际情况可能需要不同的实现
    await navigator.clipboard.writeText(text);
    successCb?.();
  } catch (error) {
    console.error('Failed to copy text to clipboard:', error);
    errorCb?.();
  }
}

const isNode =
  typeof process !== 'undefined' &&
  process.versions !== null &&
  process.versions.node !== null;

/**
 * 用于判断当前是否在浏览器环境中。
 * 首先会判断当前是否处于测试环境中（通过 process.env.NODE_ENV === 'TEST' 判断），
 * 如果是，则返回 true。否则，会进一步判断是否存在 window 对象、document 对象以及 matchMedia 方法
 * 同时通过 !isNode 判断当前不是在服务器（Node.js）环境下执行，
 * 如果都符合，则返回 true 表示当前处于浏览器环境中。
 * @returns  boolean
 */
export const isBrowser = () => {
  if (typeof process !== 'undefined' && process.env.NODE_ENV === 'TEST') {
    return true;
  }
  return (
    typeof window !== 'undefined' &&
    typeof window.document !== 'undefined' &&
    typeof window.matchMedia !== 'undefined' &&
    !isNode
  );
};
