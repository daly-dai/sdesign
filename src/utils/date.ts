import dayjs, { Dayjs } from 'dayjs';

import { RangeValueType } from '../components';

function isDayjsInstance(obj: dayjs.Dayjs) {
  // 检查对象是否存在并且不是原始类型
  if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
    // 检查对象是否有 Day.js 常见的属性或方法
    // 注意：这里只是一个示例，Day.js 的 API 可能会变化
    // 你可以根据 Day.js 的当前版本和你想要检查的具体属性/方法进行修改
    return (
      typeof obj.format === 'function' && // Day.js 实例有 format 方法
      typeof obj.toDate === 'function' && // Day.js 实例可以转换为 Date 对象
      // ... 可以添加更多 Day.js 特有的检查
      true // 如果上述条件都满足，则假设它是 Day.js 实例
    );
  }
  return false; // 如果不是对象，或者没有上述属性/方法，则不是 Day.js 实例
}

export function getDateVal(data: RangeValueType | undefined) {
  if (!data) return null;

  if (typeof data === 'string') {
    if (!dayjs(data)?.isValid()) return null;

    return dayjs(data);
  }

  if (Array.isArray(data) && data?.length) {
    return data?.map((item: string | null) => {
      if (!dayjs(item)?.isValid()) return null;

      return dayjs(item);
    });
  }

  if (typeof data === 'object' && isDayjsInstance(data as Dayjs)) {
    return data;
  }

  return null;
}
