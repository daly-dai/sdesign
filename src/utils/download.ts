/**
 * 下载二进制流
 * @param res
 * @param name
 */
export function downloadBlob(res: any, name: string) {
  const blob = new Blob([res.data]);

  const anchor = document.createElement('a');
  const href = window.URL.createObjectURL(blob); //关键点3
  anchor.setAttribute('href', href);
  // /* 关键之处：使用download属性必须要html5的页面才行 ，而且它不会刷新，文件名及扩展名均由这里控制*/
  anchor.setAttribute('download', name); //关键点4
  anchor.click();
}

/**
 * 下载链接
 * @param href
 * @param title
 */
export function downloadLink(href: string, title?: string) {
  let findTitle: string = '';

  if (!title) {
    const list = href.split('/');
    if (list.length > 0) {
      findTitle = list[list.length - 1];
    }
  }

  const a = document.createElement('a');
  a.setAttribute('href', href);
  a.setAttribute('download', title || findTitle);
  a.click();
}
