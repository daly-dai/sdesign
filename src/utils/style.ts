import { isString } from 'lodash';
import isNumber from 'lodash/isNumber';

import { isStringNumber } from './data-types';
import { debugWarn } from './error';

const SCOPE = 'utils/style';

export function addUnit(value?: string | number, defaultUnit = 'px') {
  if (!value) return '';

  if (isNumber(value) || isStringNumber(value)) {
    return `${value}${defaultUnit}`;
  }

  if (isString(value)) {
    return value;
  }

  debugWarn(SCOPE, 'binding value must be a string or number');
}
