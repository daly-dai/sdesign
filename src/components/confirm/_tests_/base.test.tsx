import { fireEvent, render } from '@testing-library/react';
import { Button } from 'antd';
import React from 'react';
import { describe, expect, it, vi } from 'vitest';

import SConfirm from '../index'; // 假设这是你的SConfirm组件的路径

import mountTest from '@dalydb/sdesign/tests/mountTest';
import { waitFakeTimer } from '@dalydb/sdesign/tests/utils';

describe('SConfirm component', () => {
  mountTest(SConfirm);

  mountTest(() => <SConfirm type="modal" />);
  it('should render  custom content', () => {
    const { getByText, container } = render(
      <SConfirm type="modal">
        <Button type="primary">Open Modal</Button>
      </SConfirm>,
    );

    expect(getByText('Open Modal')).toBeInTheDocument();
    expect(container.querySelector('.ant-btn-primary')).toBeInTheDocument();
  });

  it('should support modal type', () => {
    const { getByText } = render(
      <SConfirm type="modal">
        <Button>Open Modal</Button>
      </SConfirm>,
    );

    expect(getByText('Open Modal')).toBeInTheDocument();

    fireEvent.click(document.body.querySelectorAll('.ant-btn')[0]);

    expect(
      document.body.querySelectorAll('.ant-modal-root')[0],
    ).toMatchSnapshot();
  });

  it('should type default Popconfirm', () => {
    const { getByText, container } = render(
      <SConfirm>
        <Button className="PopconfirmBtn">Open Modal</Button>
      </SConfirm>,
    );

    expect(getByText('Open Modal')).toBeInTheDocument();

    fireEvent.click(container.querySelectorAll('.PopconfirmBtn')[0]);

    expect(
      document.body.querySelectorAll('.ant-popover-content')[0],
    ).toMatchSnapshot();
  });

  // 应该支持点击确认按钮，关闭弹窗，并调用confirm回调函数
  it('should support click confirm button and close modal', async () => {
    const handleConfirm = vi.fn();
    const handleCancel = vi.fn();

    const { getByText, container } = render(
      <SConfirm onConfirm={handleConfirm} onCancel={handleCancel}>
        <Button className="PopconfirmBtn">Open Modal</Button>
      </SConfirm>,
    );

    expect(container.querySelector('.ant-popover')).toBe(null);

    expect(getByText('Open Modal')).toBeInTheDocument();

    fireEvent.click(container.querySelectorAll('.PopconfirmBtn')[0]);

    await waitFakeTimer(100);

    expect(document.body.querySelector('.ant-popover')).not.toBeNull();
    expect(document.body.querySelector('.ant-popover')?.className).toContain(
      'ant-popover-placement-top',
    );
    expect(document.body.querySelector('.ant-popover')).toMatchSnapshot();
  });

  it('should support custom title and description', async () => {
    const { getByText, container } = render(
      <SConfirm title="自定义标题" description="自定义描述">
        <Button className="PopconfirmBtn">Open Modal</Button>
      </SConfirm>,
    );

    expect(container.querySelector('.ant-popover')).toBe(null);

    expect(getByText('Open Modal')).toBeInTheDocument();

    fireEvent.click(container.querySelectorAll('.PopconfirmBtn')[0]);

    await waitFakeTimer(100);

    expect(
      document.body.querySelector('.ant-popconfirm-title')?.textContent,
    ).toBe('自定义标题');
    expect(
      document.body.querySelector('.ant-popconfirm-description')?.textContent,
    ).toBe('自定义描述');
  });
});
