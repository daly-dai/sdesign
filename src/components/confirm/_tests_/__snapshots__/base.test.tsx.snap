// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SConfirm component > should support click confirm button and close modal 1`] = `
<div
  class="ant-popover ant-popconfirm css-dev-only-do-not-override-3rel02 css-dev-only-do-not-override-3rel02 ant-popover-placement-top"
  style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
>
  <div
    class="ant-popover-arrow"
    style="position: absolute; bottom: 0px; left: 0px;"
  />
  <div
    class="ant-popover-content"
  >
    <div
      class="ant-popover-inner"
      role="tooltip"
    >
      <div
        class="ant-popover-inner-content"
      >
        <div
          class="ant-popconfirm-inner-content"
        >
          <div
            class="ant-popconfirm-message"
          >
            <span
              class="ant-popconfirm-message-icon"
            >
              <span
                aria-label="exclamation-circle"
                class="anticon anticon-exclamation-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="exclamation-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                  />
                </svg>
              </span>
            </span>
            <div
              class="ant-popconfirm-message-text"
            >
              <div
                class="ant-popconfirm-title"
              >
                删除提示
              </div>
              <div
                class="ant-popconfirm-description"
              >
                是否确认删除该条信息
              </div>
            </div>
          </div>
          <div
            class="ant-popconfirm-buttons"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-3rel02 ant-btn-default ant-btn-sm"
              type="button"
            >
              <span>
                否
              </span>
            </button>
            <button
              class="ant-btn css-dev-only-do-not-override-3rel02 ant-btn-primary ant-btn-sm"
              type="button"
            >
              <span>
                是
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SConfirm component > should support custom title and description with type modal 1`] = `
<div>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-3rel02 ant-btn-default modalConfirmBtn"
      type="button"
    >
      <span>
        Open Modal
      </span>
    </button>
  </div>
</div>
`;

exports[`SConfirm component > should support modal type 1`] = `undefined`;

exports[`SConfirm component > should type default Popconfirm 1`] = `
<div
  class="ant-popover-content"
>
  <div
    class="ant-popover-inner"
    role="tooltip"
  >
    <div
      class="ant-popover-inner-content"
    >
      <div
        class="ant-popconfirm-inner-content"
      >
        <div
          class="ant-popconfirm-message"
        >
          <span
            class="ant-popconfirm-message-icon"
          >
            <span
              aria-label="exclamation-circle"
              class="anticon anticon-exclamation-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="exclamation-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                />
              </svg>
            </span>
          </span>
          <div
            class="ant-popconfirm-message-text"
          >
            <div
              class="ant-popconfirm-title"
            >
              删除提示
            </div>
            <div
              class="ant-popconfirm-description"
            >
              是否确认删除该条信息
            </div>
          </div>
        </div>
        <div
          class="ant-popconfirm-buttons"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-3rel02 ant-btn-default ant-btn-sm"
            type="button"
          >
            <span>
              否
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-3rel02 ant-btn-primary ant-btn-sm"
            type="button"
          >
            <span>
              是
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SConfirm component should support click confirm button and close modal 1`] = `
<div
  class="ant-popover ant-popconfirm css-dev-only-do-not-override-1r287do css-dev-only-do-not-override-1r287do ant-popover-placement-top"
  style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
>
  <div
    class="ant-popover-arrow"
    style="position: absolute; bottom: 0px; left: 0px;"
  />
  <div
    class="ant-popover-content"
  >
    <div
      class="ant-popover-inner"
      role="tooltip"
    >
      <div
        class="ant-popover-inner-content"
      >
        <div
          class="ant-popconfirm-inner-content"
        >
          <div
            class="ant-popconfirm-message"
          >
            <span
              class="ant-popconfirm-message-icon"
            >
              <span
                aria-label="exclamation-circle"
                class="anticon anticon-exclamation-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="exclamation-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                  />
                </svg>
              </span>
            </span>
            <div
              class="ant-popconfirm-message-text"
            >
              <div
                class="ant-popconfirm-title"
              >
                删除提示
              </div>
              <div
                class="ant-popconfirm-description"
              >
                是否确认删除该条信息
              </div>
            </div>
          </div>
          <div
            class="ant-popconfirm-buttons"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-1r287do ant-btn-default ant-btn-sm"
              type="button"
            >
              <span>
                否
              </span>
            </button>
            <button
              class="ant-btn css-dev-only-do-not-override-1r287do ant-btn-primary ant-btn-sm"
              type="button"
            >
              <span>
                是
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SConfirm component should support modal type 1`] = `undefined`;

exports[`SConfirm component should type default Popconfirm 1`] = `
<div
  class="ant-popover-content"
>
  <div
    class="ant-popover-inner"
    role="tooltip"
  >
    <div
      class="ant-popover-inner-content"
    >
      <div
        class="ant-popconfirm-inner-content"
      >
        <div
          class="ant-popconfirm-message"
        >
          <span
            class="ant-popconfirm-message-icon"
          >
            <span
              aria-label="exclamation-circle"
              class="anticon anticon-exclamation-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="exclamation-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                />
              </svg>
            </span>
          </span>
          <div
            class="ant-popconfirm-message-text"
          >
            <div
              class="ant-popconfirm-title"
            >
              删除提示
            </div>
            <div
              class="ant-popconfirm-description"
            >
              是否确认删除该条信息
            </div>
          </div>
        </div>
        <div
          class="ant-popconfirm-buttons"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-1r287do ant-btn-default ant-btn-sm"
            type="button"
          >
            <span>
              否
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-1r287do ant-btn-primary ant-btn-sm"
            type="button"
          >
            <span>
              是
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
